facade/src/test/java/org/jboss/pnc/facade/providers/UserProviderImplTest.java

```diff
@@ -0,0 +1,180 @@
+
+    @InjectMocks
+    private UserProviderImpl provider;
+
+    private User userMock = prepareNewUser("tada");
+
+    @Override
+    protected AbstractProvider provider() {
+        return provider;
+    }
+
+    @Override
+    protected Repository<User, Integer> repository() {
+        return repository;
+    }
+
+    @Before
+    public void setup() {
+        List<User> users = new ArrayList<>();
+        users.add(prepareNewUser("boris"));
+        users.add(prepareNewUser("theresa"));
+        users.add(prepareNewUser("david"));
+        users.add(userMock);
+
+        fillRepository(users);
+    }
+
+    @Test
+    public void testGetExistingUser() {
+        // Prepare
+        User user = repositoryList.get(0);
+
+        when(repository.queryByPredicates(any(Predicate.class))).thenAnswer(inv -> user);
+
+        // When
+        org.jboss.pnc.dto.User userDTO = provider.getOrCreateNewUser("test");
+
+        // then
+        assertThat(userDTO.getId()).isEqualTo(user.getId().toString());
+        assertThat(userDTO.getUsername()).isEqualTo(user.getUsername());
+    }
+
+    @Test
+    public void testGetNewUser() {
+
+        when(repository.queryByPredicates(any(Predicate.class))).thenAnswer(inv -> null);
+        String username = "test";
+
+        // When
+        org.jboss.pnc.dto.User userDTO = provider.getOrCreateNewUser(username);
+
+        // then
+        assertThat(userDTO.getUsername()).isEqualTo(username);
+    }
+
+    @Test
+    public void testGetAll() {
+
+        // when
+        Page<org.jboss.pnc.dto.User> all = provider.getAll(0, 10, null, null);
+
+        // then
+        assertThat(all.getContent())
+                .hasSize(4);
+    }
+
+    @Test
+    public void testGetSpecific() {
+        // when
+        org.jboss.pnc.dto.User user = provider.getSpecific(userMock.getId().toString());
+
+        // then
+        assertThat(user.getId()).isEqualTo(userMock.getId().toString());
+        assertThat(user.getUsername()).isEqualTo(userMock.getUsername());
+    }
+
+    @Test
+    public void testStoreNewUserWithIdShouldFail() {
+
+        // when
+        org.jboss.pnc.dto.User userDTO = org.jboss.pnc.dto.User.builder()
+                .id(Integer.toString(entityId++))
+                .username("john-cormack")
+                .build();
+
+        // then: can't store new user with id already set
+        assertThatThrownBy(() -> provider.store(userDTO))
+                .isInstanceOf(InvalidEntityException.class);
+    }
+
+    @Test
+    public void testStoreNewUserWithoutId() {
+
+        // when
+        org.jboss.pnc.dto.User userDTO = org.jboss.pnc.dto.User.builder()
+                .username("john-cormack")
+                .build();
+
+        org.jboss.pnc.dto.User userDTOSaved = provider.store(userDTO);
+
+        // then
+        assertThat(userDTOSaved.getId()).isNotNull().isNotBlank();
+        assertThat(userDTOSaved.getUsername()).isEqualTo(userDTO.getUsername());
```

Comment:
I think we don't support of creating users via REST.

Reply:
I believe we do for new users. If a user is not part of the PNC database, and starts to use PNC, it'll hit the REST url '/users/current', which will automatically add it to the database so that we can use it when noting who built a build config. The username is obtained from the Keycloak server.